PEP 769: Add a 'default' keyword argument to 'attrgetter' and 'itemgetter'

facundobatista opened the pull request:
Selected number: 769

This PEP is about adding  a `default` keyword argument to `attrgetter` and `itemgetter` (from the `operator` module`).

Thanks!


## Basic requirements (all PEP Types)

* [x] Read and followed [PEP 1](https://peps.python.org/1) & [PEP 12](https://peps.python.org/12)
* [x] File created from the [latest PEP template](https://github.com/python/peps/blob/main/peps/pep-0012/pep-NNNN.rst?plain=1)
* [x] PEP has next available number, & set in filename (``pep-NNNN.rst``), PR title (``PEP 123: <Title of PEP>``) and ``PEP`` header
* [x] Title clearly, accurately and concisely describes the content in 79 characters or less
* [x] Core dev/PEP editor listed as ``Author`` or ``Sponsor``, and formally confirmed their approval
* [x] ``Author``, ``Status`` (``Draft``), ``Type`` and ``Created`` headers filled out correctly
* [x] ``PEP-Delegate``, ``Topic``, ``Requires`` and ``Replaces`` headers completed if appropriate
* [x] Required sections included
    * [x] Abstract (first section)
    * [x] Copyright (last section; exact wording from template required)
* [x] Code is well-formatted (PEP 7/PEP 8) and is in [code blocks, with the right lexer names](https://peps.python.org/pep-0012/#literal-blocks) if non-Python
* [x] PEP builds with no warnings, pre-commit checks pass and content displays as intended in the rendered HTML
* [x] Authors/sponsor added to ``.github/CODEOWNERS`` for the PEP


## Standards Track requirements

* [x] PEP topic [discussed in a suitable venue](https://peps.python.org/pep-0001/#start-with-an-idea-for-python) with general agreement that a PEP is appropriate
* [ ] [Suggested sections](https://peps.python.org/pep-0012/#suggested-sections) included (unless not applicable)
    * [x] Motivation
    * [x] Rationale
    * [x] Specification
    * [x] Backwards Compatibility
    * [x] Security Implications
    * [x] How to Teach This
    * [ ] Reference Implementation
    * [x] Rejected Ideas
    * [x] Open Issues
* [x] ``Python-Version`` set to valid (pre-beta) future Python version, if relevant
* [x] Any project stated in the PEP as supporting/endorsing/benefiting from the PEP formally confirmed such
* [ ] Right before or after initial merging, [PEP discussion thread](https://peps.python.org/pep-0001/#discussing-a-pep) created and linked to in ``Discussions-To`` and ``Post-History``

<!-- readthedocs-preview pep-previews start -->
----
📚 Documentation preview 📚: https://pep-previews--4179.org.readthedocs.build/pep-0769/

<!-- readthedocs-preview pep-previews end -->

hugovk commented:
Hello!

> Selected number: 790

You can use 790 if you like, but we normally pick the next one in sequence, which currently would be 769. Would you prefer to use 769?


I've also updated the PR description to include the new PEP checklist, please could you work down it and check them off as appropriate?

Thanks!

hugovk commented:
Please could you also wrap lines to 79 columns? https://peps.python.org/pep-0012/#general

If you have prettier installed, this can do it:

```sh
prettier --prose-wrap=always --print-width=79 -w filename.rst
```

facundobatista commented:
Linters and builders fixed!

hugovk commented:
You're welcome! 

> * I've chosen 790 because the biggest one 0xxx was 0789, I didn't search for gaps. I'm fine with using 769, however I have a question: the branch is named `add-pep-790`, if I rename it (`git branch -m add-pep-769`) will we get a new PR? is that ok?

Let's go for 769 then.

If you make a new branch, you'll need to make a new PR. But the branch name isn't too important and the branch will be deleted after merge, so I'd just keep the current one.

facundobatista commented:
Cool, thanks! I renamed it to 769, then. Also updated PR title and description.

Let me know when this is ready for the discussion in the forum, and I'll open a thread there. 

Thanks!

hugovk commented on peps/pep-0769.rst:
```diff
@@ -0,0 +1,355 @@
+    except (TypeError, IndexError, KeyError):
+        value = XYZ
+
+However, this would be not as eficient as we'd want for particular cases,
```
typo:
```suggestion
However, this would be not as efficient as we'd want for particular cases,
```

hugovk commented on peps/pep-0769.rst:
```diff
@@ -0,0 +1,355 @@
+    ('bar', 'XYZ')
+
+
+About Posible Implementations
+-----------------------------
```
```suggestion
About Possible Implementations
------------------------------
```

hugovk commented on peps/pep-0769.rst:
```diff
@@ -0,0 +1,355 @@
+    try:
+        result = obj[name]
+    except (TypeError, IndexError, KeyError):
+        resutl = default
```
```suggestion
        result = default
```

hugovk commented on peps/pep-0769.rst:
```diff
@@ -0,0 +1,355 @@
+We need to define how ``itemgetter`` would behave, if just attempt to
+access the item and capture exceptions no matter which the object, or
+validate first if the object provides a ``get`` method and use it to
+retrieve the item with a default. See examples in the About Posible
+Implementations subsection before.
```
```suggestion
retrieve the item with a default. See examples in the About Possible
Implementations subsection above.
```
Or we could add a reference?

For example, add this above:


```diff
+.. _PEP 769 About Possible Implementations:
 
 About Posible Implementations
 -----------------------------
```

And then:

```suggestion
retrieve the item with a default. See examples in the `About Possible
Implementations <PEP 769 About Possible Implementations_>`__ subsection above.
```

hugovk commented on peps/pep-0769.rst:
```diff
@@ -0,0 +1,355 @@
+individual default values to multiple attributes or items. While some
+discussions considered allowing multiple defaults, the increased
+complexity and potential for confusion led to favoring a single default
+value for all cases (more about this below in Rejected Ideas).
```
We could add similar reference here if we want to hyperlink it.

hugovk commented on peps/pep-0769.rst:
```diff
@@ -0,0 +1,355 @@
+
+For the case of ``attrgetter`` is quite direct: it implies using
+``getattr`` catching a possible ``AttributeError``. So
+``attrgetter("name", default=XYZ)(obj)`` would be like:
```
So the following code is shown as code, not blockquote:

```suggestion
``attrgetter("name", default=XYZ)(obj)`` would be like::
```

hugovk commented on peps/pep-0769.rst:
```diff
@@ -0,0 +1,355 @@
+            value = XYZ
+
+Better performance, more complicated to implement and explain. This is
+the first case in the Open Issues section later.
```
This could also be a linked reference.

hugovk commented on peps/pep-0769.rst:
```diff
@@ -0,0 +1,355 @@
+    except (TypeError, IndexError, KeyError):
+        resutl = default
+
+(however see previous open issue about special case for dictionaries)
```
```suggestion
(However see previous open issue about special case for dictionaries.)
```

facundobatista commented:
Thanks! There I followed all your suggestions. I'm not 100% sure if the linked sections is fine, Github is not rendering those correctly but I understand that it may be a Github thing not supporting ReST 100% ok.

Thanks again!

hugovk commented on peps/pep-0769.rst:
```diff
@@ -0,0 +1,363 @@
+discussions considered allowing multiple defaults, the increased
+complexity and potential for confusion led to favoring a single default
+value for all cases (more about this below in `Rejected Ideas
+<PEP 769 Rejected Ideas>`__).
```
We need an extra underscore here:
```suggestion
<PEP 769 Rejected Ideas_>`__).
```

hugovk commented on peps/pep-0769.rst:
```diff
@@ -0,0 +1,363 @@
+            value = XYZ
+
+Better performance, more complicated to implement and explain. This is
+the first case in the `Open Issues <PEP 769 Open Issues>`__ section later.
```
```suggestion
the first case in the `Open Issues <PEP 769 Open Issues_>`__ section later.
```

hugovk commented on peps/pep-0769.rst:
```diff
@@ -0,0 +1,363 @@
+access the item and capture exceptions no matter which the object, or
+validate first if the object provides a ``get`` method and use it to
+retrieve the item with a default. See examples in the `About Possible
+Implementations <PEP 769 About Possible Implementations>`__ subsection
```
```suggestion
Implementations <PEP 769 About Possible Implementations_>`__ subsection
```

facundobatista commented:
Awesome, thanks! There I incorporated those fixes. Thanks for the help!

hugovk commented:
Feel free to merge it you're ready.

Then open a discussion on https://discuss.python.org/c/peps/19 and open a new PR add the link to the Post-History to the PEP.

facundobatista commented:
Thank you very much!
